Sensors adxl372 support

mhennerich opened the pull request:
This patch set adds support for the Analog Devices ADXL372 ultra-low power,
3-axis, +/-200 g MEMS accelerometer. The ADXL372 can be either connected
via a SPI or I2C interface.
Along with driver, a sample sensor application is provided which tests and
exercises the device driver in various configuration modes.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

mhennerich committed:
drivers: sensors: adxl372: Add driver for ADXL372 high-g accelerometer

This patch adds support for the Analog Devices ADXL372 ultra-low power,
3-axis, +/-200 g MEMS accelerometer. The ADXL372 can be either connected
via a SPI or I2C interface.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

mike-scott commented:
@mhennerich I'm currently working with the ADXL362 driver and I'm wondering if we can't combine this driver with that one.  Unifying the 2 drivers would ease future maintenance and provide parity for the ADXL drivers.

mhennerich commented:
Hi Mike,

I was thinking about the same thing when I started. But these two parts are very different.
The register map doesn’t match at all. If you find a register that looks similar, they are not.
The SPI transfer protocol is completely different, and the XL362 doesn’t have a I2C option.
The scales and ranges are different and features like die temperature only exits on the XL362,
while on the other side the XL372 has the max peak stuff.

Given this list the parts don’t share much in common, trying to support them in one driver defeats the purpose.

-Michael

mike-scott commented:
Ok, that makes sense.  Thanks for the explanation.

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,94 @@
+Description
+***********
+
+This sample application produces slightly different outputs based on the chosen
```
This would be a bit clearer if you used a bullet list for the different configuration modes:

    This sample application produces slightly different outputs based on the chosen
    driver configuration mode: 

    * In **Measuring Mode with trigger support**,
      the acceleration on all three axis is printed in m/s^2 at the sampling rate (ODR).
    * In **Polled Measuring Mode**, the instantaneous acceleration is polled every 2 seconds.
    * In **Max Peak Detect Mode**, the device returns only the over-threshold
      peak acceleration between two consecutive sample fetches or trigger events.
      (In most high-g applications, a single 3-axis acceleration sample at the peak
      of an impact event contains sufficient information about the event, and the
      full acceleration history is not required.)

    Instead of printing the acceleration on all three axis, the sample application
    calculates the vector magnitude (root sum squared) and displays the result
    in g's rather than in m/s^2, together with an bar graph.

mhennerich replied:
Great idea - makes it much more readable

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,94 @@
+interrupt capable GPIO. It is a requirement that **SCLK** must be connected to
+**GND** in I2C mode. Depending on the baseboard used, the **SDA** and **SCL**
+lines require Pull-Up resistors.
+Note: When sharing an SDA bus, the ADXL372 Silcon REV < 3  may prevent
```
If this is an important note, you can write this as:

    .. note::

       When sharing an SDA bus, the ADXL372 Silcon REV < 3  may prevent
       communication with other devices on that bus.

mhennerich replied:
ok great

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,94 @@
+SPI mode
+========
+
+Connect Interface: **SCLK**, **MISO**, **MOSI** and **/CS** and optionally
```
Should this be ``CS`` rather than ``/CS``?

mhennerich replied:
Ideally this would have an over-line. 
But using '/' is a common notation to describe a signal being  active low.

dbkinder replied:
Ah, OK.

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,94 @@
+   AX=     -0.98 AY=      4.90 AZ=     -3.92 (m/s^2)
+
+   <repeats endlessly>
+
```
Blank line at the end of file probably got a warning from the CI checks.

mhennerich replied:
Wasn't sure what the common practice is here. For source code files an empty line is always required.
I'll remove it then.

MaureenHelm commented on drivers/sensor/adxl372/adxl372_trigger.c:
```diff
@@ -0,0 +1,170 @@
+			ADXL372_STATUS_2_INACT(status2)) {
+			drv_data->th_handler(dev, &drv_data->th_trigger);
+		} else if (ADXL372_STATUS_2_INACT(status2) ||
+			ADXL372_STATUS_2_INACT(status2)) {
```
Why are you OR-ing `ADXL372_STATUS_2_INACT(status2)` with itself?

mhennerich replied:
Good catch - This must be ADXL372_STATUS_2_ACTIVITY.

MaureenHelm commented on samples/sensor/adxl372/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Add a copyright/license header

mhennerich replied:
ok

mhennerich commented on drivers/sensor/adxl372/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+zephyr_library()
```
Hi Maureen,

Thanks for your review!

Will do.

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+
+config ADXL372_I2C_MASTER_DEV_NAME
+	string
+	prompt "I2C master where ADXL372 is connected"
```
ok

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+	  This option is useful if one needs to manage SPI CS through a GPIO
+	  pin to by-pass the SPI controller's CS logic.
+
+config ADXL372_SPI_GPIO_CS_DRV_NAME
```
ok

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+	  In this mode the device returns only the over threshold
+	  Peak Acceleration between two consecutive sample fetches.
+
+config ADXL372_MEASUMENT_MODE
```
ups

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+config ADXL372_MEASUMENT_MODE
+	bool "Measurement Mode"
+	help
+	  Measurement mode is the default operating mode of the ADXL372.
```
If you look at an accelerometer in general. This measurement mode would be considered the default mode. I can remove the default option.

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+
+static void adxl372_accel_convert(struct sensor_value *val, s16_t value)
+{
+	s32_t micro_ms2 = value * (SENSOR_G / 160);
```
Sensor resolution is 100mg/LSB, 12-bit value needs to be right shifted by 4 or divided by 16. Overall this results in a scale of 160. I'll add some comment and make things clearer.

s32_t micro_ms2 = value * (SENSOR_G / (16 * 1000 / 100));

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+	const struct adxl372_dev_config *cfg = dev->config->config_info;
+
+#ifdef CONFIG_ADXL372_I2C
+	data->spi  = device_get_binding(cfg->i2c_port);
```
ok - I name it ->bus then.

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+	}
+	/* CPOL=0, CPHA=0, max 10MHz */
+	data->spi_cfg.operation = SPI_WORD_SET(8) | SPI_TRANSFER_MSB;
+	data->spi_cfg.frequency = 8000000;
```
Max frequency can come from the dt. The operation stuff is device specific.

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+choice
+	prompt "Accelerometer sampling frequency (ODR)"
+	default ADXL372_ODR_6400HZ if ADXL372_PEAK_DETECT_MODE
+	default ADXL372_ODR_400HZ if ADXL372_MEASUMENT_MODE
```
Sorry - I don't understand this comment. Which line?
This just sets some defaults the user typically want's to start with.

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+	  Threshold for in-activity detection.
+
+config ADXL372_ACTIVITY_TIME
+	int "Aactivity time"
```
ups

mhennerich commented on drivers/sensor/adxl372/Kconfig:
```diff
@@ -0,0 +1,320 @@
+	range 0 255
+	default 2
+	help
+	  the time that all enabled axes must be lower than the inactivity
```
ok

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+				      ADXL372_TO_I2C_REG(reg),
+				      (u8_t *) data, length);
+	} else {
+		u8_t buf[2] = {ADXL372_TO_I2C_REG(reg), *(u8_t *)data};
```
good idea

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+	if (c == ADXL372_HPF_DISABLED)
+		mask = ADXL372_POWER_CTL_HPF_DIS_MSK;
+	else
+		mask = 0;
```
ok

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+		*status2 = buf[1];
+
+	if (fifo_entries)
+		*fifo_entries = ((buf[2] & 0x3) << 8) | buf[3];
```
May consider fixing checkpatch to catch these ...

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+			      struct adxl372_xyz_accel_data *samples,
+			      u16_t cnt)
+{
+	u8_t buf[1024];
```
Reduced to 128 bytes and added a second loop, to split things into smaller junks.

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+ */
+int adxl372_service_fifo_ev(struct device *dev,
+			    struct adxl372_xyz_accel_data *fifo_data,
+				u16_t *fifo_entries)
```
Not yet - I have a second patch-set or RFC waiting but wanted to get the driver submitted first.

The FIFO buffer is an important feature in ultralow power applications in two ways:

The FIFO enables system level power savings by enabling the host processor to sleep for extended periods while the accelerometer autonomously collects data. Alternatively, using the FIFO to collect data can unburden the host while it tends to other tasks.

The FIFO can be used in a triggered mode to record all data leading up to an activity detection event, thereby providing context for the event. In the case of a system that identifies impact events, for example, the accelerometer can keep the entire system off while it stores acceleration data in its FIFO and looks for an activity event. When the impact event occurs, data collected prior to the event is frozen in the FIFO. The accelerometer can now trigger/wake the rest of the system and transfer this data to the host processor, thereby providing context for the impact event.

However to properly utilize the FIFO we need to extend the sensor_driver_api with a new channel_get method where the user specifies how many samples it can accept and then the method would return how many samples it actually returns.

MaureenHelm replied:
I agree FIFO support is a valuable feature (see #1387), but please wait to add this function (and any related helper functions) until we have some way to use it. An RFC for the sensor_driver_api would be welcome.

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+
+	do {
+		adxl372_get_status(dev, &status1, NULL, NULL);
+	} while (!(ADXL372_STATUS_1_DATA_RDY(status1)));
```
In polled mode it's mandatory to check the RDY status prior in reading the conversion registers.
In triggered mode this is already beeing checked in the trigger thread.
I'll skip this check for the triggered option.

mhennerich commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1081 @@
+		reg = ADXL372_X_THRESH_ACT_H;
+		/* Fallthrough */
+	case SENSOR_ATTR_LOWER_THRESH:
+		if (!reg) {
```
right

MaureenHelm commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1104 @@
+	} else {
+		return i2c_reg_write_byte(adxl372_data->bus, cfg->i2c_addr,
+					  ADXL372_TO_I2C_REG(reg),
+					  *(u8_t *)data);
```
This looks better with `i2c_reg_write_byte` but I still don't understand why in this particular case you assume `length` is 1. In every other case (spi read/write, i2c read), you use the passed value.

mhennerich replied:
write is slow path - we only do burst reads.
So for write - the upper layer code uses always length=1.
Does it make sense to use i2c_burst_write_addr(), knowingly we just add overhead.
Or shall I return error if length!=1? Or leave as is?

MaureenHelm commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1104 @@
+
+	do {
+		if (total > sizeof(buf)) {
+			cnt = sizeof(buf);
```
`cnt` appears to have a new meaning here, different from the function argument. Please use a new variable.

mhennerich replied:
ok

MaureenHelm commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1104 @@
+	value = abs((micro_ms2 * 10) / SENSOR_G);
+
+	if (value > 2047)
+		return -EINVAL;
```
Add braces around this line

mhennerich replied:
ok

mhennerich commented on drivers/sensor/adxl372/adxl372.h:
```diff
@@ -0,0 +1,376 @@
+
+struct adxl372_data {
+	struct device *bus;
+	struct spi_config spi_cfg;
```
This ifdef hell is ugly - but you're right

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,104 @@
+References
+**********
+
+ - ADXL372: http://www.analog.com/adxl372
```
Missed this on the first pass.  This renders better if there's no space before the list bullet (don't indent).

mhennerich replied:
I see - should be fixed now. 
Thanks!

dbkinder commented on samples/sensor/adxl372/README.rst:
```diff
@@ -0,0 +1,104 @@
+
+I2C Address:
+
+ * **0x1D**: if MISO is pulled low
```
same here, no space before the list bullet (don't indent these lines)

dbkinder reviewed (changes requested):
couple of little tweaks

mhennerich commented:
Hi All,

AFAICT all requested changes have been applied. Please let me know if there are still open issues.

MaureenHelm commented on drivers/sensor/adxl372/adxl372.c:
```diff
@@ -0,0 +1,1112 @@
+}
+
+/* Exclude until we have a API to better support this */
+#if 0
```
Please don't do this. Remove the FIFO functions from this PR then add them later when we have a way to use them.

mhennerich replied:
Removed this dead code and pushed again.

Thanks!

mhennerich commented:
recheck

mhennerich committed:
samples: sensor: adxl372: Add ADXL372 sample application

This sample application produces slightly different outputs based on
the chosen driver configuration mode. In Measuring Mode with trigger
support, the acceleration on all three axis is printed in m/s^2 at
the sampling rate (ODR). In polled Measuring Mode the instantaneous
acceleration is polled every 2 seconds. In most high-g applications,
a single (3-axis) acceleration sample at the peak of an impact event
contains sufficient information about the event, and the full
acceleration history is not required.
In this Max Peak Detect Mode the device returns only the over
threshold peak acceleration between two consecutive sample fetches or
trigger events. Instead of printing the acceleration on all three axis,
the sample application calculates the vector magnitude
(root sum squared) and displays the result in g rather than in m/s^2,
together with an bar graph.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

dbkinder reviewed (approved):
Docs LGTM, thanks.
